src/SciAssist/utils/data_utils.py

```diff
@@ -1440,3 +1439,189 @@ def get_dataloader(self, dataset, inputs_column="text", labels_column="summary")
+                use_fast=True
+            )
+        else:
+            self.tokenizer = tokenizer
+
+
+    def tokenize_and_align_labels(self, examples, inputs_column="text", labels_column="summary"):
+
+        """
+
+        Process the dataset for model input, for example, do tokenization and prepare label_ids.
+
+        Args:
+            examples (`Dataset`): { "text": [s1, s2, ...], "summary": [l1, l2, ...]}
+            inputs (`str`): The name of input column
+            labels (`str`): The name of target column
+
+        Returns:
+            `Dict`: {"input_ids": input_ids, "attention_mask": attention_mask, "labels": label_ids }
+
+        """
+
+        inputs = examples[inputs_column]
+        prompts = [ "" for i in inputs ]
+
+        # kw_instructions = ["{}"]
+        def kw(prompt, keyword):
+            # i = randint(1,10)
+            if keyword is not None and keyword != [""]:
+                return str(" | ".join(keyword)) if keyword is not None else prompt
+            return prompt
+
+        def leng(prompt, length):
+            if length is not None:
+                return prompt + ", which has less than " + str(length) + " words " if length is not None else prompt
+            return prompt
+
+        if "keywords" in examples.keys():
+            keywords = examples["keywords"]
+            if keywords is not None:
+                prompts = [ kw(prompt,keyword) for (prompt,keyword) in zip(prompts,keywords) ]
+
+        if "length" in examples.keys():
+            if examples["length"] is not None:
+                lengths = examples["length"]
+                prompts = [ leng(prompt,length) for (prompt,length) in zip(prompts,lengths)]
+
+        inputs = [ prompt + ": " + raw_text for (prompt,raw_text) in zip(prompts, inputs) ]
+
+        print(inputs[0][:200])
+        # Setup the tokenizer for inputs
+
+        model_inputs = self.tokenizer(inputs, max_length=self.max_source_length, padding="max_length", truncation=True)
+
+        # Select target column
+        if labels_column in examples.keys():
+            labels = examples[labels_column]
+            # Setup the tokenizer for targets
+            with self.tokenizer.as_target_tokenizer():
+                labels = self.tokenizer(labels, max_length=self.max_target_length, padding="max_length", truncation=True)
+                # Ignore padding in the loss
+                labels["input_ids"] = [
+                    [(l if l != self.tokenizer.pad_token_id else -100) for l in label] for label in labels["input_ids"]
+                ]
+            model_inputs["labels"] = labels["input_ids"]
+
+        #when test
+        if "id" in examples.keys():
+            model_inputs["id"] = examples["id"]
+        if "length" in examples.keys() and examples["length"] is not None:
+            if examples["length"][0] is not None:
+                model_inputs["length"] = examples["length"]
+        return model_inputs
+
+    def collator(self):
+
+        """
```

Comment:
Good sample function with descriptions. But don't understand the current description, is this collating function designed following another tool?